### PR TITLE
csi: delete stale subvolumesnapshot

### PR DESCRIPTION
**Pull request by yati1998**
this commit checks if there are any stale
volumesnapshots, if yes, it deletes the
snapshots

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -51,6 +51,7 @@ require (
 	github.com/json-iterator/go v1.1.12 // indirect
 	github.com/k8snetworkplumbingwg/network-attachment-definition-client v1.6.0 // indirect
 	github.com/kube-object-storage/lib-bucket-provisioner v0.0.0-20221122204822-d1a8c34382f1 // indirect
+	github.com/kubernetes-csi/external-snapshotter/client/v7 v7.0.0
```
Comment (Madhu-1) [inline go.mod:54]: use v8 client.

File: pkg/filesystem/subvolume.go [hunk 2]
```diff
@@ -131,6 +139,36 @@ func getK8sRefSubvolume(ctx context.Context, clientsets *k8sutil.Clientsets) map
...
+
+	snapshotHandles := make(map[string]snapshotInfo)
+	for _, snap := range snapList.Items {
+		if snap.Status.SnapshotHandle != nil {
+			// get the snaps id from snapshot handle
+			splitSnapshotHandle := strings.SplitAfterN(*snap.Status.SnapshotHandle, "-", 6)
+			if len(splitSnapshotHandle) < 6 {
+				return nil
+			}
+			snapshotHandleId := splitSnapshotHandle[len(splitSnapshotHandle)-1]
```
Comment (Madhu-1) [inline pkg/filesystem/subvolume.go:162]: does this also take care of https://github.com/ceph/ceph-csi/blob/d5849a4801fa9f3383cefc4f96e18fcad420dc80/examples/cephfs/snapshotclass.yaml#L19C5-L19C23

File: pkg/filesystem/subvolume.go [hunk 3]
```diff
@@ -131,6 +139,36 @@ func getK8sRefSubvolume(ctx context.Context, clientsets *k8sutil.Clientsets) map
...
+// getk8sRefSnapshotHandle returns the snapshothandle for k8s ref of the volume snapshots
+func getK8sRefSnapshotHandle(ctx context.Context, clientsets *k8sutil.Clientsets) map[string]snapshotInfo {
+
+	Snap, err := snapclient.NewForConfig(clientsets.KubeConfig)
+	if err != nil {
+		logging.Fatal(err)
+	}
+	snapList, err := Snap.VolumeSnapshotContents().List(ctx, v1.ListOptions{})
+	if err != nil {
+		logging.Fatal(fmt.Errorf("Error fetching snapshots: %v\n", err))
```
Comment (Madhu-1) [inline pkg/filesystem/subvolume.go:151]: volumesnapshotcontent

File: pkg/filesystem/subvolume.go [hunk 4]
```diff
@@ -131,6 +139,36 @@ func getK8sRefSubvolume(ctx context.Context, clientsets *k8sutil.Clientsets) map
...
+	}
+
+	snapshotHandles := make(map[string]snapshotInfo)
+	for _, snap := range snapList.Items {
+		if snap.Status.SnapshotHandle != nil {
```
Comment (Madhu-1) [inline pkg/filesystem/subvolume.go:156]: Status is a pointer you need nil check for it as well

File: pkg/filesystem/subvolume.go [hunk 5]
```diff
@@ -263,13 +302,36 @@ func checkSnapshot(ctx context.Context, clientsets *k8sutil.Clientsets, operator
...
 	snap := unMarshaljson(snapList)
+	// check for stale subvolume snapshot
+	// we have the list of snapshothandleid's from the
+	// volumesnapshotcontent. Looking up for snapid in it
+	// will confirm if we have stale snapshot or not.
+	for _, s := range snap {
+		snapId := getSnapId(s.Name)
+		// lookup for the snapid in the k8s snapshot handle list
+		_, ok := snapshotHandles[snapId]
+		logging.Info("%s", snapId)
```
Comment (Madhu-1) [inline pkg/filesystem/subvolume.go:313]: this logging doesnt seems to be useful can you please check?

**Review by Madhu-1 — CHANGES_REQUESTED** [review]
@yati1998 what about the omap cleanup?

File: pkg/filesystem/subvolume.go (hunk 2 above)
Comment (yati1998) [reply pkg/filesystem/subvolume.go:162]: yes, it just compares the snap-id ignore the prefix which matches with snapshothandle.

File: go.mod [hunk 6]
```diff
@@ -1,19 +1,22 @@
 module github.com/rook/kubectl-rook-ceph
 
-go 1.21
+go 1.22.0
```
Comment (travisn) [inline go.mod:3]: Let's make the toolchain update with a separate PR. @subhamkrai Want to take care of that?

File: pkg/filesystem/subvolume.go [hunk 7]
```diff
@@ -131,6 +139,36 @@ func getK8sRefSubvolume(ctx context.Context, clientsets *k8sutil.Clientsets) map
 	return subvolumeNames
 }
 
+// getk8sRefSnapshotHandle returns the snapshothandle for k8s ref of the volume snapshots
+func getK8sRefSnapshotHandle(ctx context.Context, clientsets *k8sutil.Clientsets) map[string]snapshotInfo {
+
+	Snap, err := snapclient.NewForConfig(clientsets.KubeConfig)
```
Comment (travisn) [inline pkg/filesystem/subvolume.go:145]: Local vars generally start with lower case (camelCase). How about this var name?
```suggestion
	snapConfig, err := snapclient.NewForConfig(clientsets.KubeConfig)
```

File: pkg/filesystem/subvolume.go [hunk 8]
```diff
@@ -131,6 +139,36 @@ func getK8sRefSubvolume(ctx context.Context, clientsets *k8sutil.Clientsets) map
...
+	snapList, err := Snap.VolumeSnapshotContents().List(ctx, v1.ListOptions{})
+	if err != nil {
+		logging.Fatal(fmt.Errorf("Error fetching volumesnapshotcontents: %v\n", err))
+	}
+
+	snapshotHandles := make(map[string]snapshotInfo)
+	for _, snap := range snapList.Items {
+		if snap.Status != nil && snap.Status.SnapshotHandle != nil {
+			// get the snaps id from snapshot handle
+			splitSnapshotHandle := strings.SplitAfterN(*snap.Status.SnapshotHandle, "-", 6)
```
Comment (travisn) [inline pkg/filesystem/subvolume.go:158]: Could you add a unit test for this string parsing? It will help show the need, as well as prevent regressions in the future. If it's simpler for unit testing, you might want to refactor this parsing into a separate helper method.

File: pkg/filesystem/subvolume.go [hunk 9]
```diff
@@ -296,6 +348,19 @@ func unMarshaljson(list string) []fsStruct {
 	return unmarshal
 }
 
+// deleteSnapshot deletes the subvolume snapshot
+func deleteSnapshot(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, fs, subvol, svg, snap string) {
```
Comment (travisn) [inline pkg/filesystem/subvolume.go:352]: parameters are generally `camelCase`
```suggestion
func deleteSnapshot(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespace, cephClusterNamespace, fs, subvol, svg, snap string) {
```

File: pkg/filesystem/subvolume.go [hunk 10]
```diff
@@ -480,3 +607,19 @@ func getOmapVal(subVol string) (string, string) {
 
 	return omapval, subvolId
 }
+
+// func getSnapOmapVal is used to get the omapval from the given snapshot
+// omapval is of format csi.snap.427774b4-340b-11ed-8d66-0242ac110005
+// which is similar to volume name csi-snap-427774b4-340b-11ed-8d66-0242ac110005
+// hence, replacing 'csi-vol-' to 'csi.snap.'
+func getSnapOmapVal(snap string) (string, string) {
```
Comment (travisn) [inline pkg/filesystem/subvolume.go:615]: Please add a unit test for this method

File: pkg/filesystem/subvolume.go [hunk 11]
```diff
@@ -407,6 +507,33 @@ func getOmapKey(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNam
...
+	args := []string{"getomapval", snapomapval, "csi.snapname", "-p", poolName, "--namespace", "csi", "/dev/stdout"}
+	cmd := "rados"
+	snapshotcontentname, err := runCommand(ctx, clientsets, OperatorNamespace, CephClusterNamespace, cmd, args)
+	if err != nil || snapshotcontentname == "" {
+		logging.Info("No snapshot content found for snapshot %s: %s", snap, err)
+		return ""
+	}
+	// omap key is for format csi.volume.pvc-fca205e5-8788-4132-979c-e210c0133182
+	// hence, attaching pvname to required prefix.
+	snapomapkey := "csi.snap." + snapshotcontentname
```
Comment (travisn) [inline pkg/filesystem/subvolume.go:532]: From the comment up two lines, should this prefix be "csi.volume"? Or else I don't understand the comment and why it's "csi.snap".

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/filesystem/subvolume.go (hunk 11 above)
Comment (yati1998) [reply pkg/filesystem/subvolume.go:532]: my bad, thanks for noticing.

File: pkg/filesystem/subvolume.go [hunk 12]
```diff
@@ -131,6 +139,47 @@ func getK8sRefSubvolume(ctx context.Context, clientsets *k8sutil.Clientsets) map
...
+// SnapshotHandle: 0001-0009-rook-ceph-0000000000000001-17b95621-
+// 58e8-4676-bc6a-39e928f19d23
+// SnapshotHandleId: 17b95621-58e8-4676-bc6a-39e928f19d23
+func getSnapshotHandleId(snapshotHandle string) string {
+	// get the snaps id from snapshot handle
+	splitSnapshotHandle := strings.SplitAfterN(snapshotHandle, "-", 6)
+	if len(splitSnapshotHandle) < 6 {
+		return ""
+	}
+	snapshotHandleId := splitSnapshotHandle[len(splitSnapshotHandle)-1]
```
Comment (travisn) [inline pkg/filesystem/subvolume.go:178]: nit: return it directly instead of declaring a var
```suggestion
	return splitSnapshotHandle[len(splitSnapshotHandle)-1]
```

File: pkg/filesystem/subvolume.go [hunk 13]
```diff
@@ -195,11 +244,11 @@ func listCephFSSubvolumes(ctx context.Context, clientsets *k8sutil.Clientsets, o
 					// check the state of the stale subvolume
 					// if it is snapshot-retained then skip listing it.
 					if state == "snapshot-retained" {
```
Comment (travisn) [inline pkg/filesystem/subvolume.go:246]: Can we also use this var here?
```suggestion
					if state == snapshotRetained {
```

File: pkg/filesystem/subvolume.go [hunk 14]
```diff
@@ -407,6 +518,33 @@ func getOmapKey(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNam
 	return omapkey
 }
 
+// getSnapOmapKey gets the omap key and value details for a given snapshot.
+// Deletion of omap object required the snapshotName which is of format
+// csi.snap.snapid.
+// similarly to delete of omap key requires csi.snap.ompakey, where
+// omapkey is the snapshotcontent name which is extracted the omap object.
+func getSnapOmapKey(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, snap, fs string) string {
```
Comment (travisn) [inline pkg/filesystem/subvolume.go:526]:
```suggestion
func getSnapOmapKey(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespace, cephClusterNamespace, snap, fs string) string {
```

**Review by travisn — CHANGES_REQUESTED** [review]

**Comment by yati1998** [comment]
@travisn @Madhu-1 please do review the PR, addressed all the changes.

File: pkg/filesystem/subvolume_test.go [hunk 15]
```diff
@@ -59,3 +59,72 @@ func TestGetOmapVal(t *testing.T) {
...
+			val:  "58e8-4676-bc6a-39e928f19d23",
+		},
+		{
+			name: "rook-427774b440b11ed8d660242ac11000",
+			val:  "",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if val := getSnapshotHandleId(tt.name); val != tt.val {
```
Comment (travisn) [inline pkg/filesystem/subvolume_test.go:125]: This is a bit simpler for failing the test if the values are not equal. It will format the expected and actual message for you.
```suggestion
			if val := getSnapshotHandleId(tt.name); val != tt.val {
				assert.Equal(t, val, tt.val)
			}
```

**Review by travisn — COMMENTED** [review]
LGTM, just a small test suggestion

**Comment by yati1998** [comment]
@subhamkrai do you have any idea why is the ci failing, it seems it is failing due to some plugin test. Can you please check?

File: go.sum [hunk 16]
```diff
@@ -521,6 +521,10 @@ github.com/kr/text v0.2.0/go.mod h1:eLer722TekiGuMkidMxC/pM04lWEeraHUUmBw8l2grE=
 github.com/kube-object-storage/lib-bucket-provisioner v0.0.0-20221122204822-d1a8c34382f1 h1:dQEHhTfi+bSIOSViQrKY9PqJvZenD6tFz+3lPzux58o=
 github.com/kube-object-storage/lib-bucket-provisioner v0.0.0-20221122204822-d1a8c34382f1/go.mod h1:my+EVjOJLeQ9lUR9uVkxRvNNkhO2saSGIgzV8GZT9HY=
 github.com/kubernetes-csi/external-snapshotter/client/v4 v4.0.0/go.mod h1:YBCo4DoEeDndqvAn6eeu0vWM7QdXmHEeI9cFWplmBys=
+github.com/kubernetes-csi/external-snapshotter/client/v7 v7.0.0 h1:j3YK74myEQRxR/srciTpOrm221SAvz6J5OVWbyfeXFo=
+github.com/kubernetes-csi/external-snapshotter/client/v7 v7.0.0/go.mod h1:FlyYFe32mPxKEPaRXKNxfX576d1AoCzstYDoOOnyMA4=
```
Comment (Madhu-1) [inline go.sum:525]: can you update other files to use v8 client as well, both v4 and v7 entry might need to go away from this file

File: pkg/filesystem/subvolume.go [hunk 17]
```diff
@@ -40,6 +41,11 @@ type subVolumeInfo struct {
 	state string
 }
 
+type snapshotInfo struct {
+	volumehandle   string
```
Comment (Madhu-1) [inline pkg/filesystem/subvolume.go:45]: `volumehandle` to `volumeHandle` and same for snapshotHandle

File: pkg/filesystem/subvolume.go [hunk 18]
```diff
@@ -131,6 +139,47 @@ func getK8sRefSubvolume(ctx context.Context, clientsets *k8sutil.Clientsets) map
...
+
+	snapConfig, err := snapclient.NewForConfig(clientsets.KubeConfig)
+	if err != nil {
+		logging.Fatal(err)
+	}
+	snapList, err := snapConfig.VolumeSnapshotContents().List(ctx, v1.ListOptions{})
+	if err != nil {
+		logging.Fatal(fmt.Errorf("Error fetching volumesnapshotcontents: %v\n", err))
+	}
+
```
Comment (Madhu-1) [inline pkg/filesystem/subvolume.go:153]: can we skip all other volumesnapshotcontent except cephfs driver related?

File: pkg/filesystem/subvolume.go [hunk 19]
```diff
@@ -379,6 +455,41 @@ func deleteOmapForSubvolume(ctx context.Context, clientsets *k8sutil.Clientsets,
...
+		// remove omap object.
+		_, err := runCommand(ctx, clientsets, OperatorNamespace, CephClusterNamespace, cmd, args)
+		if err != nil {
+			logging.Fatal(err, "failed to remove omap object for snapshot %q", snap)
+		}
+		logging.Info("omap object:%q deleted", snapomapval)
+
+	}
+	if snapomapkey != "" {
+		cmd := "rados"
```
Comment (Madhu-1) [inline pkg/filesystem/subvolume.go:480]: move this one to line 466 and remove 468

File: pkg/filesystem/subvolume.go [hunk 20]
```diff
@@ -407,6 +518,33 @@ func getOmapKey(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNam
...
+	}
+	snapomapval, _ := getSnapOmapVal(snap)
+
+	args := []string{"getomapval", snapomapval, "csi.snapname", "-p", poolName, "--namespace", "csi", "/dev/stdout"}
+	cmd := "rados"
+	snapshotcontentname, err := runCommand(ctx, clientsets, operatorNamespace, cephClusterNamespace, cmd, args)
+	if err != nil || snapshotcontentname == "" {
+		logging.Info("No snapshot content found for snapshot %s: %s", snap, err)
+		return ""
+	}
```
Comment (Madhu-1) [inline pkg/filesystem/subvolume.go:540]: am not sure this check is correct, we can consider the error !=nil as not found right?

File: pkg/filesystem/subvolume.go [hunk 21]
```diff
@@ -480,3 +618,19 @@ func getOmapVal(subVol string) (string, string) {
 
 	return omapval, subvolId
 }
+
+// func getSnapOmapVal is used to get the omapval from the given snapshot
+// omapval is of format csi.snap.427774b4-340b-11ed-8d66-0242ac110005
+// which is similar to volume name csi-snap-427774b4-340b-11ed-8d66-0242ac110005
+// hence, replacing 'csi-vol-' to 'csi.snap.'
```
Comment (Madhu-1) [inline pkg/filesystem/subvolume.go:625]: do you mean replacing `csi-snap-` to `csi.snap.`?

Comment (yati1998) [reply pkg/filesystem/subvolume.go:625]: yes

File: pkg/filesystem/subvolume.go (hunk 20 above)
Comment (yati1998) [reply pkg/filesystem/subvolume.go:540]: yes correct.

File: go.mod [hunk 22]
```diff
@@ -5,6 +5,7 @@ go 1.21
 require (
 	github.com/fatih/color v1.17.0
 	github.com/golang/mock v1.6.0
+	github.com/kubernetes-csi/external-snapshotter/client/v8 v8.0.0
```
Comment (subhamkrai) [inline go.mod:8]: do your changes require this change?

Comment (yati1998) [reply go.mod:8]: yes

**Review by travisn — CHANGES_REQUESTED** [review]
The CI is failing:
```
go: github.com/kubernetes-csi/external-snapshotter/client/v8@v8.0.0 requires go >= 1.22.0 (running go 1.21.11)
make: *** [Makefile:19: build] Error 1
```

**Comment by yati1998** [comment]
> The CI is failing:
> 
> ```
> go: github.com/kubernetes-csi/external-snapshotter/client/v8@v8.0.0 requires go >= 1.22.0 (running go 1.21.11)
> make: *** [Makefile:19: build] Error 1
> ```

yes, checked that, we need to update the go.mod here itself.

**Comment by subhamkrai** [comment]
> > The CI is failing:
> > ```
> > go: github.com/kubernetes-csi/external-snapshotter/client/v8@v8.0.0 requires go >= 1.22.0 (running go 1.21.11)
> > make: *** [Makefile:19: build] Error 1
> > ```
> 
> yes, checked that, we need to update the go.mod here itself or send another PR and merge it first?



> > The CI is failing:
> > ```
> > go: github.com/kubernetes-csi/external-snapshotter/client/v8@v8.0.0 requires go >= 1.22.0 (running go 1.21.11)
> > make: *** [Makefile:19: build] Error 1
> > ```
> 
> yes, checked that, we need to update the go.mod here itself or send another PR and merge it first?

@yati1998 I created this issue https://github.com/rook/kubectl-rook-ceph/issues/309 please rebase your PR once that issue is closed.

File: pkg/filesystem/subvolume.go [hunk 23]
```diff
@@ -131,6 +139,48 @@ func getK8sRefSubvolume(ctx context.Context, clientsets *k8sutil.Clientsets) map
...
+	}
+	snapList, err := snapConfig.VolumeSnapshotContents().List(ctx, v1.ListOptions{})
+	if err != nil {
+		logging.Fatal(fmt.Errorf("Error fetching volumesnapshotcontents: %v\n", err))
+	}
+
+	snapshotHandles := make(map[string]snapshotInfo)
+	for _, snap := range snapList.Items {
+		driverName := snap.Spec.Driver
+		if snap.Status != nil && snap.Status.SnapshotHandle != nil && strings.Contains(driverName, "cephfS") {
```
Comment (Madhu-1) [inline pkg/filesystem/subvolume.go:157]: `cephfS` to `cephfs`

File: pkg/filesystem/subvolume.go (hunk 20 above)
Comment (Madhu-1) [reply pkg/filesystem/subvolume.go:540]: This looks like a wrong assumption to consider `err!=nil` as not found and continue , can you please check this again

**Comment by yati1998** [comment]
>>This looks like a wrong assumption to consider err!=nil as not found and continue , can you please check this again

@Madhu-1 I think this is fine, either there is no omap or some error running the command, anywhich ways we are printing the error too. we do the same for pv also, looks fine to me.

File: pkg/filesystem/subvolume.go (hunk 20 above)
Comment (Madhu-1) [reply pkg/filesystem/subvolume.go:540]: Having this assumption will leave a stale entry in the omap, can you please check, if there is any error when running this command it should be logged as Fatal not just info and return empty struct which will skip omap deletion operation.

Comment (yati1998) [reply pkg/filesystem/subvolume.go]: okay checked that, added the required change

Comment (yati1998) [reply pkg/filesystem/subvolume.go]: @Madhu-1 please do review again.

File: pkg/filesystem/subvolume.go [hunk 24]
```diff
@@ -407,6 +518,36 @@ func getOmapKey(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNam
...
+	poolName, err := getMetadataPoolName(ctx, clientsets, operatorNamespace, cephClusterNamespace, fs)
+	if err != nil || poolName == "" {
+		logging.Fatal(fmt.Errorf("pool name not found: %q", err))
+	}
+	snapomapval, _ := getSnapOmapVal(snap)
+
+	args := []string{"getomapval", snapomapval, "csi.snapname", "-p", poolName, "--namespace", "csi", "/dev/stdout"}
+	cmd := "rados"
+	snapshotcontentname, err := runCommand(ctx, clientsets, operatorNamespace, cephClusterNamespace, cmd, args)
+	if snapshotcontentname == "" {
```
Comment (Madhu-1) [inline pkg/filesystem/subvolume.go:537]: this should also have `err==nil` check to work as expected or else if its err!=nil we will have it snapshotcontentname set to empty string

Comment (yati1998) [reply pkg/filesystem/subvolume.go:537]: oh yeah, got it.

**Comment by travisn** [comment]
@yati1998 Could you look at the CI? Not sure at a glance what the failure is.

**Comment by travisn** [comment]
@yati1998 Please rebase it to resolve the merge conflicts

**Comment by yati1998** [comment]
> @yati1998 Please rebase it to resolve the merge conflicts

can you please review now?

**Review by travisn — APPROVED** [review]
